File: src/main/java/dev/jason/project/spring/vc_server/domain/Device.java
package dev.jason.project.spring.vc_server.domain;

import java.time.LocalDateTime;

public record Device(String name, OS os, String version, String fcmToken, LocalDateTime lastUsed) {

    public enum OS {
        Android // currently only android is supported. may add support to other oses in the future.
    }

    @Override
    public boolean equals(Object other) {
        if (!(other instanceof Device)) return false;
        return ((Device) other).fcmToken.equals(fcmToken);
    }
}

File: core/src/main/java/dev/jason/app/compose/vaultchat/core/domain/Device.kt
package dev.jason.app.compose.vaultchat.core.domain

import android.content.Context
import android.content.res.Configuration
import android.os.Build

data class Device(
    val name: String,
    val type: Type,
    val os: Os,
    val version: String,
    val token: String,
) {
    enum class Os {
        Android // currently only android is supported
    }

    enum class Type {
        Mobile, Tablet
    }

    companion object {
        fun getCurrentDevice(context: Context, token: String): Device {
            return Device(
                name = "${Build.MANUFACTURER} ${Build.MODEL}",
                os = Os.Android,
                version = "${Build.VERSION.RELEASE}",
                token = token,
                type = getDeviceType(context)
            )
        }

        private fun getDeviceType(context: Context): Type {
            val screenLayout = context.resources.configuration.screenLayout
            val mask = Configuration.SCREENLAYOUT_SIZE_MASK
            val isTablet = (screenLayout and mask) >= Configuration.SCREENLAYOUT_SIZE_LARGE
            return if (isTablet) {
                Type.Tablet
            } else {
                Type.Mobile
            }
        }
    }
}